docs/rfc/rc-005-event-system.rst

```diff
@@ -0,0 +1,121 @@
+=====================
+RFC 005: Event System
+=====================
+
+Changelog
+---------
+
+- 2021-09-17: Initial Draft (@tychoish)
+
+Abstract
+--------
+
+The event system within Tendermint, which supports a lot of core
+functionality, also represents a major infrastructural liability. As part of
+our upcoming review of the RPC interfaces and our ongoing thoughts about
+stability and performance, as well as the preparation for Tendermint 1.0, we
+should revisit the design and implementation of the event system. This
+document discusses both the current state of the system 
+
+Background
+----------
+
+Current State of Events
+~~~~~~~~~~~~~~~~~~~~~~~
+
+The event system makes it possible for clients, both internal and external,
+to receive notifications after a series of given events, such as new blocks,
+new transactions, validator set changes, as well as intermediate events during
+consensus. Because the event system is very cross cutting, the behavior and
+performance of the event publication and subscription system has huge impacts
+for all of Tendermint.
+
+The subscription service is exposed over the RPC interface, but also powers
+the indexing (e.g. to an external database,) and is the mechanism by which
+`BroadcastTxCommit` is able to wait for transactions to land in a block.
+
+The current pubsub mechanism relies on a couple of buffered channels,
```

Comment:
Specifically: The subscriber engine maintains one as an inbound work queue, and there is another channel per subscriber-client (which may or may not be buffered) for publication, plus at least one other unbuffered(?) channel for waiting on shutdown. A slow subscriber is a DoS on the engine's work loop, which doesn't block new subscribers from being queued, but does prevent them from being registered, and can block publication. It's not quite deadlock, but it may be molasses-lock 🙂

Reply:
Yuuuuup. 

I'm going to leave this comment here for posterity.